server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -423,11 +437,35 @@ protected void setMinSeqNoToKeep(long seqNo) {
         this.minSeqNoToKeep = seqNo;
     }
 
+    @Override
+    protected Releasable drainSync() {
+        try {
+            if (syncPermit.tryAcquire(SYNC_PERMIT, 1, TimeUnit.MINUTES)) {
+                logger.info("All inflight remote translog syncs finished and further syncs paused");
+                return Releasables.releaseOnce(() -> {
+                    syncPermit.release(SYNC_PERMIT);
+                    assert syncPermit.availablePermits() == SYNC_PERMIT : "Available permits is " + syncPermit.availablePermits();
+                    logger.info("Resumed remote translog sync back on relocation failure");
+                });
+            } else {
+                throw new TimeoutException("Timeout while acquiring all permits");
+            }
+        } catch (TimeoutException | InterruptedException e) {
+            throw new RuntimeException("Failed to acquire all permits", e);
+        }
+    }
+
     @Override
     public void trimUnreferencedReaders() throws IOException {
         // clean up local translog files and updates readers
         super.trimUnreferencedReaders();
 
+        // This is to ensure that after the permits are acquired during primary relocation, there are no further modification on remote
+        // store.
+        if (syncPermit.availablePermits() != SYNC_PERMIT) {
+            return;
+        }
+
```

Comment:
This can also impact other flows while like when a concurrent upload is happening

Reply:
Used the same approach used in ReleasableRefreshListener. This would not be impacted by concurrent upload now.